### ISSUE TITLE
write_kafka plugin only working with -f flag.

### ISSUE THREAD

**Issue by utkarshcmu** [collectd/collectd#996]
write_kafka plugin is not working without -f flag. -f is for "Don't fork to the background.", this plugin is not working in background.

**Comment by mfournier** [comment]
Thanks for reporting this @utkarshcmu. I'm a bit surprised. Can you define "not working" more precisely please? Thanks!

**Comment by utkarshcmu** [comment]
When I am using the "-f" flag for running CollectD Daemon with "write_kafka" plugin enabled and configured properly, the CollectD Daemon is able to write to Kafka. But when I am not using "-f" flag, I do not see any writes happening to Kafka keeping the same configuration and running the same CollectD Daemon.

**Comment by mfournier** [comment]
OK, can you try enabling the syslog plugin to see of there's any error message? Also maybe try using tcpdump/tshark/ngrep to see if there's any network traffic between collectd and kafka.

**Comment by mfournier** [comment]
I can confirm this issue. Setting up a basic kafka service following https://kafka.apache.org/documentation.html#quickstart, building collectd from current master (against librdkafka 0.8.5) and simply uncommenting the sample configuration block for write_kafka is sufficient to reproduce the problem.

Using kafka's `bin/kafka-console-consumer.sh` we can see collectd data flow in when the daemon is run in the foreground, but nothing when it's run in the background. In both cases no error show up in collectd logs, only plugin: `plugin_write: Writing values via write_kafka/test.`

One difference that I noticed:

* when running in the foreground:
```
$ lsof -p $(pidof collectd) -nn
lsof: WARNING: can't stat() ext4 file system /vm/docker/aufs
      Output information may be incomplete.
COMMAND    PID USER   FD   TYPE             DEVICE SIZE/OFF     NODE NAME
collectd 22224 marc  cwd    DIR              254,0     4096   107956 /opt/collectd/var/lib/collectd
collectd 22224 marc  rtd    DIR              254,0     4096        2 /
collectd 22224 marc  txt    REG              254,0   586608     3796 /opt/collectd/sbin/collectd
collectd 22224 marc  mem    REG              254,0    47712  1364038 /lib/x86_64-linux-gnu/libnss_files-2.19.so
collectd 22224 marc  mem    REG              254,0    31784  1364069 /lib/x86_64-linux-gnu/librt-2.19.so
collectd 22224 marc  mem    REG              254,0   109144  1315765 /lib/x86_64-linux-gnu/libz.so.1.2.8
collectd 22224 marc  mem    REG              254,0   199728   133997 /usr/lib/x86_64-linux-gnu/librdkafka.so.1
collectd 22224 marc  mem    REG              254,0    87568    88803 /opt/collectd/lib/collectd/write_kafka.so
collectd 22224 marc  mem    REG              254,0    35592    10421 /opt/collectd/lib/collectd/battery.so
collectd 22224 marc  mem    REG              254,0    17352    88751 /opt/collectd/lib/collectd/syslog.so
collectd 22224 marc  mem    REG              254,0  1729984  1344863 /lib/x86_64-linux-gnu/libc-2.19.so
collectd 22224 marc  mem    REG              254,0    14664  1363897 /lib/x86_64-linux-gnu/libdl-2.19.so
collectd 22224 marc  mem    REG              254,0    39392   133966 /usr/lib/x86_64-linux-gnu/libltdl.so.7.3.0
collectd 22224 marc  mem    REG              254,0   137440  1311567 /lib/x86_64-linux-gnu/libpthread-2.19.so
collectd 22224 marc  mem    REG              254,0  1051056  1363911 /lib/x86_64-linux-gnu/libm-2.19.so
collectd 22224 marc  mem    REG              254,0   140928  1344850 /lib/x86_64-linux-gnu/ld-2.19.so
collectd 22224 marc    0u   CHR             136,14      0t0       17 /dev/pts/14
collectd 22224 marc    1u   CHR             136,14      0t0       17 /dev/pts/14
collectd 22224 marc    2u   CHR             136,14      0t0       17 /dev/pts/14
collectd 22224 marc    3u  unix 0xffff8800019f7780      0t0 24838747 socket
collectd 22224 marc    5u  IPv4           24840431      0t0      TCP 127.0.0.1:35191->127.0.0.1:9092 (ESTABLISHED)
collectd 22224 marc    7u  IPv4           24837849      0t0      TCP 127.0.0.1:43820->127.0.1.1:9092 (ESTABLISHED)
```
And when I kill collectd, kafka's log display:
```
[2015-04-12 10:38:44,894] INFO Closing socket connection to /127.0.0.1. (kafka.network.Processor)
[2015-04-12 10:38:44,894] INFO Closing socket connection to /127.0.0.1. (kafka.network.Processor)
```

* and when running in the background:
```
$ lsof -p $(pidof collectd) -nn
lsof: WARNING: can't stat() ext4 file system /vm/docker/aufs
      Output information may be incomplete.
COMMAND    PID USER   FD   TYPE             DEVICE SIZE/OFF     NODE NAME
collectd 22192 marc  cwd    DIR              254,0     4096   107956 /opt/collectd/var/lib/collectd
collectd 22192 marc  rtd    DIR              254,0     4096        2 /
collectd 22192 marc  txt    REG              254,0   586608     3796 /opt/collectd/sbin/collectd
collectd 22192 marc  mem    REG              254,0    47712  1364038 /lib/x86_64-linux-gnu/libnss_files-2.19.so
collectd 22192 marc  mem    REG              254,0    31784  1364069 /lib/x86_64-linux-gnu/librt-2.19.so
collectd 22192 marc  mem    REG              254,0   109144  1315765 /lib/x86_64-linux-gnu/libz.so.1.2.8
collectd 22192 marc  mem    REG              254,0   199728   133997 /usr/lib/x86_64-linux-gnu/librdkafka.so.1
collectd 22192 marc  mem    REG              254,0    87568    88803 /opt/collectd/lib/collectd/write_kafka.so
collectd 22192 marc  mem    REG              254,0    35592    10421 /opt/collectd/lib/collectd/battery.so
collectd 22192 marc  mem    REG              254,0    17352    88751 /opt/collectd/lib/collectd/syslog.so
collectd 22192 marc  mem    REG              254,0  1729984  1344863 /lib/x86_64-linux-gnu/libc-2.19.so
collectd 22192 marc  mem    REG              254,0    14664  1363897 /lib/x86_64-linux-gnu/libdl-2.19.so
collectd 22192 marc  mem    REG              254,0    39392   133966 /usr/lib/x86_64-linux-gnu/libltdl.so.7.3.0
collectd 22192 marc  mem    REG              254,0   137440  1311567 /lib/x86_64-linux-gnu/libpthread-2.19.so
collectd 22192 marc  mem    REG              254,0  1051056  1363911 /lib/x86_64-linux-gnu/libm-2.19.so
collectd 22192 marc  mem    REG              254,0   140928  1344850 /lib/x86_64-linux-gnu/ld-2.19.so
collectd 22192 marc    0u   CHR                1,3      0t0     1028 /dev/null
collectd 22192 marc    1u   CHR                1,3      0t0     1028 /dev/null
collectd 22192 marc    2u   CHR                1,3      0t0     1028 /dev/null
collectd 22192 marc    3u  unix 0xffff8800b2f71440      0t0 24838544 socket
collectd 22192 marc    5u  IPv4           24840274      0t0      TCP 127.0.0.1:35190->127.0.0.1:9092 (ESTABLISHED)
```
(nb: only one connection to kafka)
and the following gets printed in kafka's logs when collectd exits:
```
[2015-04-12 10:38:57,817] ERROR Closing socket for /127.0.0.1 because of error (kafka.network.Processor)
java.io.IOException: Connection reset by peer
        at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
        at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
        at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
        at sun.nio.ch.IOUtil.read(IOUtil.java:197)
        at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:379)
        at kafka.utils.Utils$.read(Utils.scala:380)
        at kafka.network.BoundedByteBufferReceive.readFrom(BoundedByteBufferReceive.scala:54)
        at kafka.network.Processor.read(SocketServer.scala:444)
        at kafka.network.Processor.run(SocketServer.scala:340)
        at java.lang.Thread.run(Thread.java:745)
```

I guess we'll need @pyr to take a look :-/

**Comment by mfournier** [comment]
This problem also occurs with librdkafka 0.8.6 and current librdkafka master.

**Comment by arielshaqed** [comment]
I am unfamiliar with librdkafka, and did not find mention of this in the documentation. But: for a very similar problem, @yosih showed me that the config plugin method is called before forking, while the write plugin method is called after forking.
Here's what I think is the problem. The config method calls rd_kafka_new(), which allocates a new thread:
https://github.com/edenhill/librdkafka/blob/49673696b14790735f6dbfd2f230851c8d1ef3d9/src/rdkafka.c#L1291
Now (unless -f is specified) collectd fork()s and loses this thread, and presumably rd_kafka won't work.

Solutions I know of: put something in an init method (which needs to know all the different configurations), or do only configuration in the config method and initialize rd_kafka on the first call to the write plugin.

Does this sound relevant?

**Comment by ciomaire** [comment]
had a go at the suggestion from @arielshaqed above "do only configuration in the config method and initialize rd_kafka on the first call to the write plugin."

With this patch the write_kafka is successfully producing mesages to my kafka cluster in daemon mode.

Hope it not a problem pasting the patch directly in this ticket.

```
diff --git a/src/write_kafka.c b/src/write_kafka.c
index ccc4784..fb98970 100644
--- a/src/write_kafka.c
+++ b/src/write_kafka.c
@@ -38,6 +38,7 @@
 #include <librdkafka/rdkafka.h>
 #include <pthread.h>
 #include <zlib.h>
+#include <errno.h>

 struct kafka_topic_context {
 #define KAFKA_FORMAT_JSON        0
@@ -48,6 +49,7 @@ struct kafka_topic_context {
     _Bool                        store_rates;
     rd_kafka_topic_conf_t       *conf;
     rd_kafka_topic_t            *topic;
+    rd_kafka_conf_t             *kafka_conf;
     rd_kafka_t                  *kafka;
     int                          has_key;
     u_int32_t                    key;
@@ -55,8 +57,10 @@ struct kafka_topic_context {
     char                        *postfix;
     char                         escape_char;
     char                        *topic_name;
+    pthread_mutex_t            lock;
 };

+static int kafka_handle(struct kafka_topic_context *);
 static int kafka_write(const data_set_t *, const value_list_t *, user_data_t *);
 static int32_t kafka_partition(const rd_kafka_topic_t *, const void *, size_t,
                                int32_t, void *, void *);
@@ -85,6 +89,56 @@ static int32_t kafka_partition(const rd_kafka_topic_t *rkt,
     return target;
 }   

+static int kafka_handle(struct kafka_topic_context *ctx) /* {{{ */
+{
+    char                         errbuf[1024];
+    rd_kafka_conf_t             *conf;
+    rd_kafka_topic_conf_t       *topic_conf;
+
+    if (ctx->kafka != NULL && ctx->topic != NULL)
+        return(0);
+
+    if (ctx->kafka == NULL) {
+        if ((conf = rd_kafka_conf_dup(ctx->kafka_conf)) == NULL) {
+            ERROR("write_kafka plugin: cannot duplicate kafka config");
+            return(1);
+        }
+
+        if ((ctx->kafka = rd_kafka_new(RD_KAFKA_PRODUCER, conf,
+                                    errbuf, sizeof(errbuf))) == NULL) {
+               ERROR("write_kafka plugin: cannot create kafka handle.");
+               return 1;   
+        }
+
+       rd_kafka_conf_destroy(ctx->kafka_conf);
+       INFO ("write_kafka plugin: created KAFKA handle : %s", rd_kafka_name(ctx->kafka));
+
+#ifdef HAVE_LIBRDKAFKA_LOGGER
+       rd_kafka_set_logger(ctx->kafka, kafka_log);
+#endif
+    }
+
+    if (ctx->topic == NULL ) {
+       if ((topic_conf = rd_kafka_topic_conf_dup(ctx->conf)) == NULL) {
+            ERROR("write_kafka plugin: cannot duplicate kafka topic config");
+            return 1;
+       }
+
+       if ((ctx->topic = rd_kafka_topic_new(ctx->kafka, ctx->topic_name,
+                                                       topic_conf)) == NULL) {
+               ERROR("write_kafka plugin: cannot create topic : %s\n",
+                       rd_kafka_err2str(rd_kafka_errno2err(errno)));
+               return errno;
+       }
+
+       rd_kafka_topic_conf_destroy(ctx->conf);
+       INFO ("write_kafka plugin: handle created for topic : %s", rd_kafka_topic_name(ctx->topic));
+    }
+
+    return(0);
+
+} /* }}} int kafka_handle */
+
 static int kafka_write(const data_set_t *ds, /* {{{ */
              const value_list_t *vl,
              user_data_t *ud)
@@ -100,6 +154,12 @@ static int kafka_write(const data_set_t *ds, /* {{{ */
     if ((ds == NULL) || (vl == NULL) || (ctx == NULL))
         return EINVAL;

+    pthread_mutex_lock (&ctx->lock);
+    status = kafka_handle(ctx);
+    pthread_mutex_unlock (&ctx->lock);
+    if( status != 0 )
+        return status;
+
     bzero(buffer, sizeof(buffer));

     switch (ctx->format) { 
@@ -164,6 +224,10 @@ static void kafka_topic_context_free(void *p) /* {{{ */
         rd_kafka_topic_destroy(ctx->topic);
     if (ctx->conf != NULL) 
         rd_kafka_topic_conf_destroy(ctx->conf);
+    if (ctx->kafka_conf != NULL)
+        rd_kafka_conf_destroy(ctx->kafka_conf);
+    if (ctx->kafka != NULL)
+        rd_kafka_destroy(ctx->kafka);

     sfree(ctx);
 } /* }}} void kafka_topic_context_free */
@@ -193,19 +257,15 @@ static void kafka_config_topic(rd_kafka_conf_t *conf, oconfig_item_t *ci) /* {{{
 #ifdef HAVE_LIBRDKAFKA_LOG_CB
     rd_kafka_conf_set_log_cb(conf, kafka_log);
 #endif
-    if ((tctx->kafka = rd_kafka_new(RD_KAFKA_PRODUCER, conf,
-                                    errbuf, sizeof(errbuf))) == NULL) {
+
+    if ((tctx->kafka_conf = rd_kafka_conf_dup(conf)) == NULL) {
         sfree(tctx);
-        ERROR("write_kafka plugin: cannot create kafka handle.");
+        ERROR("write_kafka plugin: cannot allocate memory for kafka config");
         return;
     }
-#ifdef HAVE_LIBRDKAFKA_LOGGER
-    rd_kafka_set_logger(tctx->kafka, kafka_log);
-#endif
-    conf = NULL;

     if ((tctx->conf = rd_kafka_topic_conf_new()) == NULL) {
-        rd_kafka_destroy(tctx->kafka);
+        rd_kafka_conf_destroy(tctx->kafka_conf);
         sfree(tctx);
         ERROR ("write_kafka plugin: cannot create topic configuration.");
         return;
@@ -327,13 +387,6 @@ static void kafka_config_topic(rd_kafka_conf_t *conf, oconfig_item_t *ci) /* {{{
     rd_kafka_topic_conf_set_partitioner_cb(tctx->conf, kafka_partition);
     rd_kafka_topic_conf_set_opaque(tctx->conf, tctx);

-    if ((tctx->topic = rd_kafka_topic_new(tctx->kafka, tctx->topic_name,
-                                       tctx->conf)) == NULL) {
-        ERROR("write_kafka plugin: cannot create topic.");
-        goto errout;
-    }
-    tctx->conf = NULL;
-
     ssnprintf(callback_name, sizeof(callback_name),
               "write_kafka/%s", tctx->topic_name);

@@ -347,14 +400,13 @@ static void kafka_config_topic(rd_kafka_conf_t *conf, oconfig_item_t *ci) /* {{{
                                callback_name, status);
         goto errout;
     }
+
+    pthread_mutex_init (&tctx->lock, /* attr = */ NULL);
+
     return;
  errout:
     if (conf != NULL)
         rd_kafka_conf_destroy(conf);
-    if (tctx->kafka != NULL)
-        rd_kafka_destroy(tctx->kafka);
-    if (tctx->topic != NULL)
-        rd_kafka_topic_destroy(tctx->topic);
     if (tctx->topic_name != NULL)
         free(tctx->topic_name);
     if (tctx->conf != NULL)
@@ -367,7 +419,6 @@ static int kafka_config(oconfig_item_t *ci) /* {{{ */
        int                          i;
        oconfig_item_t              *child;
     rd_kafka_conf_t             *conf;
-    rd_kafka_conf_t             *cloned;
     rd_kafka_conf_res_t          ret;
     char                         errbuf[1024];

@@ -375,16 +426,11 @@ static int kafka_config(oconfig_item_t *ci) /* {{{ */
         WARNING("cannot allocate kafka configuration.");
         return -1;
     }
-
        for (i = 0; i < ci->children_num; i++)  {
                child = &ci->children[i];

                if (strcasecmp("Topic", child->key) == 0) {
-            if ((cloned = rd_kafka_conf_dup(conf)) == NULL) {
-                WARNING("write_kafka plugin: cannot allocate memory for kafka config");
-                goto errout;
-            }
-                       kafka_config_topic (cloned, child);
+                       kafka_config_topic (conf, child);
                } else if (strcasecmp(child->key, "Property") == 0) {
                        char *key = NULL;
                        char *val = NULL;
```

**Comment by mfournier** [comment]
Hey, thanks so much for investigating this !

@arielshaqed your hypothesis make a lot of sense to me. It wouldn't be the first threading-related issue with collectd & external libraries I've heared of.

@ciomaire I tried the patch yesterday evening and it looks like it solves the problem :clap: I noticed a double-free error when stopping collectd if running in the foreground though. And valgrind mentions quite a lot of "possibly lost" memory blocks (*but* it seems to be in librdkafka and there were already some before the patch).

In any case, I want to wait for @pyr to have a chance of taking a look at this (I'm aware he's at a conference these days). Would it be possible to check this double-free error and make a proper pull-request meanwhile ? Thanks a lot !
